### PR TITLE
Do not deny logins in `isMFARequired`

### LINKED ISSUE #7652: Logins denied by RBAC rules no longer generate audit log entries
In 6.0.1, login attempts when denied by RBAC rules generated audit log entries.

Example in 6.0.1 when principal is missing from certificate:
```bash
$ tsh ssh node
ERROR: access denied to tim connecting to node on cluster foo.example.com
```

...and the associated audit log entry:
```json
{
  "addr.local": "127.0.0.1:3022",
  "addr.remote": "192.168.10.1:60180",
  "cluster_name": "",
  "code": "T3007W",
  "ei": 0,
  "error": "ssh: principal \"tim\" not in the set of valid principals for given certificate: [\"-teleport-nologin-a58cfc06-fd14-4952-83bf-995169799e49\"]",
  "event": "auth",
  "login": "tim",
  "success": false,
  "time": "2021-07-22T21:32:54.286Z",
  "uid": "c811a4a4-44ac-4d4f-8963-2e6a686ad614",
  "user": "tim2"
}
```

Similarly, if denied by node labels, the user sees this:
```bash
$ tsh ssh monitor
ERROR: access denied to tim connecting to node on cluster foo.example.com
```

... and an audit log entry is generated:
```json
{
  "addr.local": "127.0.0.1:3022",
  "addr.remote": "192.168.10.1:60145",
  "cluster_name": "",
  "code": "T3007W",
  "ei": 0,
  "error": "user tim2@foo.example.com is not authorized to login as tim@foo.example.com: access to server denied",
  "event": "auth",
  "login": "tim",
  "success": false,
  "time": "2021-07-22T21:31:48.917Z",
  "uid": "415609bf-e510-4c6c-b1e5-e1e660a68346",
  "user": "tim2"
}
```

Now in 7.0.0-beta.3, the user sees the following message in both cases:
```bash
$ tsh ssh node
ERROR: node "Node:\"node\" Login:\"tim\" " not found
```

In both cases, the proxy server logs the attempt as an authorization failure but no audit log is generated. Port forwarding attempts denied by RBAC rules do still generate audit log events in v7.0.0-beta.3.

v6.2.3 appears to behave the same as v7.0.0.

### PR DESCRIPTION
**Pull request by timothyb89**
This changes `isMFARequired` to return `false` rather than an error (e.g. not found) when a user is not authorized to access a particular resource. This defers the authorization denial until such time as the client actually attempts to connect.

Before Teleport v6.2, clients would immediately attempt to connect to nodes, and if denied by RBAC rules, would print an "access denied" error.

As of Teleport v6.2, clients first call `isMFARequired` and, if unauthorized, abort the connection and print a "node not found" error. As no connection is ever attempted, no audit log entry is generated.

Fixes #7652

### CODE REVIEW

File: lib/auth/auth.go [hunk 1]
```diff
@@ -2377,15 +2377,18 @@ func (a *Server) isMFARequired(ctx context.Context, checker services.AccessCheck
 	if !errors.Is(noMFAAccessErr, services.ErrSessionMFARequired) {
 		if trace.IsAccessDenied(noMFAAccessErr) {
 			log.Infof("Access to resource denied: %v", noMFAAccessErr)
```
Comment (fspmarshall) [inline lib/auth/auth.go:2379]: I think this log message will now be confusing, so lets remove it.

**Review by fspmarshall — COMMENTED** [review]
Another bit that I think should be fixed as part of this PR... node checking logic currently looks like this:
```golang
for _, n := range matches {
    err := checker.CheckAccessToServer(t.Node.Login, n, services.AccessMFAParams{AlwaysRequired: false, Verified: false})
    if err != nil {
        noMFAAccessErr = err
        break
    }
}
```

That seems wrong to me.  I think we should skip errors that aren't `ErrSessionMFARequired` rather than breaking on the first error.  The scenario I'm imagining is one where two nodes match, the first one cannot be accessed by the user, and the second one requires MFA.  In that case, the current logic would break on a non-mfa error, resulting in `isMFARequired(...)` returning `false` when it aught to return `true` (since, of the matching nodes, the one the user is allowed to see/access does actually require MFA).

File: lib/auth/auth.go [hunk 2]
```diff
@@ -2312,9 +2311,13 @@ func (a *Server) isMFARequired(ctx context.Context, checker services.AccessCheck
...
 		for _, n := range matches {
 			err := checker.CheckAccessToServer(t.Node.Login, n, services.AccessMFAParams{AlwaysRequired: false, Verified: false})
-			if err != nil {
+			if err == nil {
+				continue
+			} else if errors.Is(err, services.ErrSessionMFARequired) {
 				noMFAAccessErr = err
 				break
+			} else {
+				log.Debugf("ignoring unrelated error while checking access to server: %v", err)
```
Comment (fspmarshall) [inline lib/auth/auth.go:2320]: Probably unnecessary, any error here should re-emerge during the real server access attempt.

**Review by fspmarshall — APPROVED** [review]

File: lib/auth/auth.go (hunk 2 above)
Comment (timothyb89) [reply lib/auth/auth.go]: Fair enough, done

**Review by xacrimon — APPROVED** [review]

File: lib/auth/auth.go [hunk 3]
```diff
@@ -2312,7 +2311,11 @@ func (a *Server) isMFARequired(ctx context.Context, checker services.AccessCheck
...
 			err := checker.CheckAccessToServer(t.Node.Login, n, services.AccessMFAParams{AlwaysRequired: false, Verified: false})
-			if err != nil {
+
+			// Ignore other errors; they'll be caught on the real access attempt.
+			if err == nil {
+				continue
+			} else if errors.Is(err, services.ErrSessionMFARequired) {
 				noMFAAccessErr = err
 				break
 			}
```
Comment (Joerger) [inline lib/auth/auth.go:2321]: nit:
```suggestion
			if err != nil && errors.Is(err, services.ErrSessionMFARequired) {
				noMFAAccessErr = err
				break
			}
```

**Review by Joerger — APPROVED** [review]

File: lib/auth/grpcserver_test.go [hunk 4]
```diff
@@ -929,6 +930,95 @@ func TestIsMFARequired(t *testing.T) {
...
+	resp, err = cl.IsMFARequired(ctx, &proto.IsMFARequiredRequest{
+		Target: &proto.IsMFARequiredRequest_Node{Node: &proto.NodeLogin{
+			Login: "bob",
+			Node:  "node1",
+		}},
+	})
+
+	// When unauthorized, expect a silent `false`.
+	require.NoError(t, err)
+	require.Equal(t, false, resp.Required)
```
Comment (r0mant) [inline lib/auth/grpcserver_test.go:1019]: `require.False(t, resp.Required)`

File: lib/auth/grpcserver_test.go [hunk 5]
```diff
@@ -929,6 +930,95 @@ func TestIsMFARequired(t *testing.T) {
...
+	// Call the endpoint for an authorized login. The user is only authorized
+	// for the 2nd node, but should still be asked for MFA.
+	resp, err := cl.IsMFARequired(ctx, &proto.IsMFARequiredRequest{
+		Target: &proto.IsMFARequiredRequest_Node{Node: &proto.NodeLogin{
+			Login: "alice",
+			Node:  "node1",
+		}},
+	})
+	require.NoError(t, err)
+	require.Equal(t, true, resp.Required)
```
[the remark this answers is not in the document]
Comment (timothyb89) [reply lib/auth/grpcserver_test.go:1007]: Whoops, good catch.

File: lib/auth/auth.go [hunk 6]
```diff
@@ -2376,16 +2377,17 @@ func (a *Server) isMFARequired(ctx context.Context, checker services.AccessCheck
...
-			// just in case.
-			if notFoundErr == nil {
-				notFoundErr = trace.NotFound("target resource not found")
-			}
-			// Mask access denied errors to prevent resource name oracles.
-			return nil, trace.Wrap(notFoundErr)
+			// Mask the access denied errors by returning false to prevent
+			// resource name oracles. Auth will be denied (and generate an
+			// audit log entry) when the client attempts to connect.
+			return &proto.IsMFARequiredResponse{Required: false}, nil
```
[the remark this answers is not in the document]
Comment (timothyb89) [reply lib/auth/auth.go:2383]: That's much more clear, thanks!

**Review by r0mant — APPROVED** [review]

**Pull request merged**